### ISSUE TITLE
Path improvements

### ISSUE THREAD

**Issue by easbar** [graphhopper/graphhopper#1566]
There are a few things about the `Path` class that could be improved:

- When we calculate a route between points that are not connected we get an 'empty' path, e.g. via `AbstractRoutingAlgorithm#createEmptyPath`. We can find out that there was no connection by calling `Path#isFound`, but I think this is too easy to forget. There is no exception when we call `getDistance` or `getTime` on the empty path (both return 0 contrary to weight which returns `Double.MAX_VALUE`, which is rather confusing).
I think there should either be an exception when calling `getWeight/Distance/Time` on a path that was not found or they should all return the maximum value. (The same goes for `calcNodes` and `calcPoints` probably).

- When a path is not found `Path#extract` is a no-op for the two subclasses `PathBidirRef` and `PathNative`, but `Path` yields a null pointer exception. There should probably be an exception if extract is called on an 'empty' path ?

- For routing queries with from=to the edge-based versions of the algorithms return maximum weight, but zero distance and zero time. I think the weight should also be zero in this case ? This seems like a bug more than a possible improvement.

- There is no easy way to find out that a connection was not found in `PathWrapper`. There is the errors list, but one would have to check if the error contained in it is instanceof `ConnectionNotFoundError` (not so nice). There is also the `impossible` flag, but this is only used in gtfs so far. Not sure if it was intended for this purpose, otherwise I think there should be something like `PathWrapper#connectionFound()`.

- minor: `Path#toString` only shows the distance and edgeIds, not sure why but for debugging it would be helpful to also have at least isFound, weight and time.

**Comment by karussell** [comment]
Related to #1110

**Comment by easbar** [comment]
> Related to #1110

Ah ok thanks for the pointer, I see there are a few things that could be improved. 

Speaking of the first point on my list should we throw an exception for `getTime/Dist/Weight/Nodes/Points` if `isFound()` is false ? Or return default values (`Long.max, Double.max, Double.max, emptyList, emptyList`) ? I do not think I have a preference, but think it should not be that if `isFound` is false `getWeight` returns max while `getDistance` and `getTime` return 0. This should be independent from further questions regarding the path extraction (?).

**Comment by karussell** [comment]
Yes, preferable throwing an exception (if we do not find important situations that require ignoring this case and returning "defaults")

**Comment by michaz** [comment]
> There is also the impossible flag, but this is only used in gtfs so far.

No. An impossible path is actually a path. It has a finite weight, duration, everything.. It just has the problem that you can't use it because the train is late, or cancelled.

It only makes sense with multi-criteria routing, where "being possible" is a criterion. You want the best possible path (in order to use it) and the best path overall, even if impossible (so that you know there is a problem with the train you were probably going to take.)

In road routing, an analogy would be an unexpectedly closed bridge or something.
